lib/common/src/Dom/Document.php

```diff
@@ -0,0 +1,1341 @@
+     * @param string $encoding Optional. The encoding of the document as part of the XML declaration.
+     */
+    public function __construct($version = '', $encoding = null)
+    {
+        $this->originalEncoding = (string)$encoding ?: self::UNKNOWN_ENCODING;
+        parent::__construct($version ?: '1.0', self::AMP_ENCODING);
+    }
+
+    /**
+     * Named constructor to provide convenient way of transforming HTML into DOM.
+     *
+     * @param string $html     HTML to turn into a DOM.
+     * @param string $encoding Optional. Encoding of the provided HTML string.
+     * @return Document|false DOM generated from provided HTML, or false if the transformation failed.
+     */
+    public static function fromHtml($html, $encoding = null)
+    {
+        $dom = new self('', $encoding);
+
+        if (! $dom->loadHTML($html)) {
+            return false;
+        }
+
+        return $dom;
+    }
+
+    /**
+     * Named constructor to provide convenient way of transforming a HTML fragment into DOM.
+     *
+     * The difference to Document::fromHtml() is that fragments are not normalized as to their structure.
+     *
+     * @param string $html     HTML to turn into a DOM.
+     * @param string $encoding Optional. Encoding of the provided HTML string.
+     * @return Document|false DOM generated from provided HTML, or false if the transformation failed.
+     */
+    public static function fromHtmlFragment($html, $encoding = null)
+    {
+        $dom = new self('', $encoding);
+
+        if (! $dom->loadHTMLFragment($html)) {
+            return false;
+        }
+
+        return $dom;
+    }
+
+    /**
+     * Named constructor to provide convenient way of retrieving the DOM from a node.
+     *
+     * @param DOMNode $node Node to retrieve the DOM from. This is being modified by reference (!).
+     * @return Document DOM generated from provided HTML, or false if the transformation failed.
+     */
+    public static function fromNode(DOMNode &$node)
+    {
+        $root = $node->ownerDocument;
+
+        // If the node is the document itself, ownerDocument returns null.
+        if (null === $root) {
+            $root = $node;
+        }
+
+        if ($root instanceof Document) {
+            return $root;
+        }
+
+        $dom = new self();
+
+        // We replace the $node by reference, to make sure the next lines of code will
+        // work as expected with the new document.
+        // Otherwise $dom and $node would refer to two different DOMDocuments.
+        $node = $dom->importNode($root->documentElement ?: $root, true);
+        $dom->appendChild($node);
+
+        return $dom;
+    }
+
+    /**
+     * Add a tweak to modify the HTML output after it was generated from the DOM.
+     *
+     * Uses the class name as array key to deduplicate tweaks.
+     *
+     * @param Tweak $tweak Callable that takes the produced HTML string and tweaks it as needed.
+     */
+    public function addTweak(Tweak $tweak)
```

Comment:
Actually, thinking about this some more... There is only one Tweak currently, right? And isn't the SVG source tweak temporary since switching to single-quoted attributes will eliminate the need for it?  Maybe better to just add the logic as a method like `restoreSelfClosingTags` and just call it explicitly in the `saveHTMLFragment` method. This will avoid creating a whole new construct.

Reply:
The problem were not the quotes, we were using double quotes all along. The problem is that the output gets mangled on `saveHTML()` automatically by `DOMDocument`, like turning spaces into `%20`.

Reply:
Btw, the entire `Tweak` code portion to fix the SVG issue is not finished yet. This was the initial code I produced while on a plane, but I couldn't manage to get the validation tool to run over the plane's network, so I don't even know yet whether the code even works. I didn't continue work on it at that point.

Reply:
Also, I don't feel good about adding "fixing logic" to `Dom\Document` about something it shouldn't be aware of. We'll probably face this more in the future, so having a mechanism for transformers to add this kind of "post-saveHTML()" manipulation is valuable and cleaner.

Reply:
Yes, those all sound better. As described above, `Tweak` was a first draft of the code only, with a temporary name.

How about a more generic and symmetric mechanism that we could also use to clean up the `Dom\Document` implementation itself:
`PreLoadHtmlFilter` & `PostSaveHtmlFilter` ?

Reply:
I'm concerned about allowing it to be pluggable. If we open an arbitrary way for site owners to manipulate the HTML, it will get abused and we won't be able to guarantee AMP-compatibility.

Reply:
Site owners already have thousands of arbitrary ways of manipulating the HTML.

This, more importantly though, is not an arbitrary, but a controlled way of manipulating the HTML. We are collecting an array of classes to do the manipulations, and these need to fulfill interface and type requirements. Therefore, we can for example include the array of manipulations into consideration when defining cache validity.

People will manipulate and misuse no matter what, after all their main priority is to get the result they need for their site, whether we provide it out of the box or not. Trying to prevent that will never work, it is preferable to rather provide a controlled way to do so.

Reply:
> Also, I don't feel good about adding "fixing logic" to `Dom\Document` about something it shouldn't be aware of. We'll probably face this more in the future, so having a mechanism for transformers to add this kind of "post-saveHTML()" manipulation is valuable and cleaner.

But our `Document` here is not just a `Dom\Document`, it is an `Amp\Dom\Document`: it is AMP-specific. So if `DOMDocument::saveHTML()` is doing something that resulting in invalid AMP, it seems logical to me to bake the logic inside of the class. We already do this for Mustache templates, noscript elements, script templates (as of #4333), self-closing tags, etc. In fact, the problem being solved here is almost exactly the same as is being addressed by `\Amp\Dom\Document::replaceMustacheTemplateTokens()`, where the URL is getting serialized incorrectly in AMP.

I'd rather put it into `Document` for now and then extract it into a separate construct when/if we have more examples that don't make sense. Having a single instance of a `Tweak` doesn't seem to warrant having this construct in place, with only the future promise of there being more such instances.